nemo_aligner/data/nlp/datasets.py

```diff
@@ -429,3 +430,144 @@ def __getitem__(self, idx):
             "labels": label_tensor,
         }
         return output
+
+
+class SteerLM2Dataset(GPTSFTChatDataset):
+    def get_prompt(self, system_turn, prompt_turns):
+        prompt = f"{self.special_tokens['system_turn_start']}System{self.special_tokens['end_of_name']}"
+        prompt += f"{system_turn}{self.special_tokens['end_of_turn']}"
+        for turn in prompt_turns:
+            prompt += f"{self.special_tokens['turn_start']}{turn['from']}{self.special_tokens['end_of_name']}"
+            prompt += f"{turn['value']}{self.special_tokens['end_of_turn']}"
+        return prompt
+
+    def _process_example(self, example):
+        """
+        Create an example by concatenating text and answer.
+        Truncation is carried out when needed, but it is performed only on the prompt side.
+        BOS, EOS, and SEP, are added if specified.
+        """
+        assert len(example["prompt_turns"]) % 2 == 1, "Number of prompt turns should be odd"
+        prompt = self.get_prompt(example["system"], example["prompt_turns"])
+        batched_token_ids = []
+        batched_masks = []
+        response_from = example["responses"][0]["from"]
+        assert [item["from"] for item in example["responses"]] == [response_from] * len(
+            example["responses"]
+        ), "All responses should be from the same person"
+        prompt += f"{self.special_tokens['turn_start']}{response_from}{self.special_tokens['end_of_name']}"
+        if "label" in example and example["label"] is not None:
+            prompt += f"{self.special_tokens['label_start']}{example['label']}{self.special_tokens['end_of_turn']}"
+        prompt_tokens = self.tokenizer.text_to_ids(prompt)
+        num_prompt_tokens = len(prompt_tokens)
+        batch_size = len(example["responses"])
+        logws = []
+        logqs = []
+        for item in example["responses"]:
+            full_text = prompt
+            full_text += item["value"] + self.special_tokens["end_of_turn"] + self.special_tokens["turn_start"]
+            token_ids = self.tokenizer.text_to_ids(full_text)
+            masks = [0] * num_prompt_tokens + [1] * (len(token_ids) - num_prompt_tokens)
+            logqs.append(item["log(Q(y|a,x))"])
+            logw = item["log(P(a|x,y))"] + item["log(P(y|x))"] - item["log(Q(y|a,x))"]
+            # for numerical stability
+            logws.append(logw)
+            # logw = logw - logw.max()
+            # w = np.exp(logw)
+            # ws.append(w)
+            batched_token_ids.append(token_ids)
+            batched_masks.append(masks)
+        logws = np.array(logws)
+        # for numerical stability
+        logws = logws - logws.max()
+        ws = np.exp(logws)
+        # noramlized weights
+        ws = ws / ws.sum()
```

Comment:
super nit:

in scipy or pytorch they actually already do this internally for stability so i think it's safe to just call softmax

[pytorch](https://github.com/pytorch/pytorch/blob/11f2d8e823efa8508d1f2198429b95b1e9007222/aten/src/ATen/native/cuda/PersistentSoftmax.cuh#L157)

[scipy softmax](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.softmax.html)